skrub/_dispatch.py

```diff
@@ -0,0 +1,194 @@
+In the same way, we can define an implementation for polars even if it is not
+installed.
+
+>>> @my_function.specialize("polars")
+... def _my_function_polars(df, other_arg=1.0):
+...     print("calling polars implementation")
+
+Now when we call ``my_function`` with a dataframe, the correct implementation is
+chosen based on the type of the first argument.
+
+>>> import pandas as pd
+>>> df = pd.DataFrame(dict(a=[3, 4]))
+>>> my_function(df)
+calling pandas implementation
+
+Inside a specialized implementation, it is safe to import the corresponding
+module or to call methods on the first argument that only exist for this
+specific type.
+
+
+>>> @my_function.specialize("polars")
+... def _my_function_polars(df, other_arg=1.0):
+...     print("calling polars implementation")
+...     # we know df is a polars DataFrame, LazyFrame or Series (see how to
+...     # specialize for only one of those types below)
+...     df.shape             # ok
+...     df.to_pandas()       # ok
+...     import polars as pl  # ok
+
+
+Outside such specialization, skrub code **must not** directly access methods of
+a container, as this would rely on backend-specific behavior:
+
+
+>>> def my_other_function(df):
+...     # not ok: will fail if df is a polars DataFrame
+...     df.values
+...     # not ok: only happens to work by chance because both polars and pandas use
+...     # the same name with the same signature and semantics, but could fail for
+...     # other backends added in the future
+...     df.shape
+...     # ok
+...     import skrub._dataframe as ns
+...     ns.shape(df)
+
+
+We can inspect which implementations were registered in ``my_function.registry``.
+
+It is also possible to register a specialization specifically for dataframes or
+for columns:
+
+>>> @my_function.specialize("pandas", "Column")
+... def _my_function_pandas_column(df, other_arg=1.0):
+...     print("calling pandas Series implementation")
+...     print("               ^^^^^^               ")
+
+>>> my_function(pd.Series([0]))
+calling pandas Series implementation
+               ^^^^^^
+>>> my_function(pd.DataFrame())
+calling pandas implementation
+
+Note that the last registered implementation wins -- there is no magic to
+prioritize specializations based on how specific they are.
+
+>>> @my_function.specialize("pandas")
+... def _override_all_pandas_specializations(df, other_arg=1.0):
+...     print("Any pandas (Series or DataFrame)")
+
+>>> my_function(pd.Series([0]))
+Any pandas (Series or DataFrame)
+>>> my_function(pd.DataFrame())
+Any pandas (Series or DataFrame)
+
+Importantly, single dispatch is done based on the type of the
+**first argument**. This means our generic function should be written in such a way
+that the first argument is the dataframe or column that will determine which
+implementation to use. For example the following would not work:
+
+>>> @dispatch
+... def sample(n_rows, df):
```

Comment:
I assume that we could raise if the first argument does not live in the namespace of the library or something like that.